Use ZCash compressed point format

selfissued opened the pull request:
Fixes #30 

The examples will also need to be updated.

Cc: @dwaite @emlun @ve7jtb

selfissued committed:
Use ZCash compressed point format

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -71,21 +71,19 @@ The following definitions apply to the pairing-friendly elliptic curves known as
-- The parameter "y" MUST be present whose value represents the y coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
+- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.
+- The parameter "d" MUST be present for private key representations with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.
```
Neither BBS or BLS currently define encoding for the secret key. BBS also exclusively uses big-endian representations otherwise, while today `d` is little-endian (to match the Ed25519/Ed448 native encoding). BBS needs to represent points in its protocol, but not secrets.

The updated text also misrepresents the value of `d` to be a compressed point rather than a scalar.

I would suggest preserving the existing `d` parameter text here and in COSE

selfissued replied:
OK - I'll apply your suggestion later this morning

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -71,21 +71,19 @@ The following definitions apply to the pairing-friendly elliptic curves known as
-- The parameter "y" MUST be present whose value represents the y coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
+- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.
+- The parameter "d" MUST be present for private key representations with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.
```
```suggestion
- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
```

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -71,21 +71,19 @@ The following definitions apply to the pairing-friendly elliptic curves known as
-- The parameter "y" (-3) Must be present whose value represents the y coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "d" (-4) MUST be present for private key representations whose value MUST contain the little-endian representation of the private key. This parameter MUST NOT be present for public keys.
+- The parameter "x" (-2) MUST be present with its value being the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.
+- The parameter "d" (-4) MUST be present for private key representations with its value being the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.
```
```suggestion
- The parameter "d" (-4) MUST be present for private key representations whose value MUST contain the little-endian representation of the private key. This parameter MUST NOT be present for public keys.
```

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -210,75 +208,67 @@ E'(GF(p^8))
-  "y": "BVkkrVEib-P_FMPHNtqxJymP3pV-H8fCdvPkoWInpFfM9tViyqD8JAmwDf64zU2h
-BV_vvCQ632ScAooEExXuz1IeQH9D2o-uY_dAjZ37YHuRMEyzh8Tq-90JHQvicOqx",
-  "d": "hR6HfxlTwcjMGST5wYnkGiJvuVnpUPbvXSGsvwjJhUM"
+  "d": "TBD THIS IS NOW WRONG. REPLACE IT. svwjJhUM"
```
```suggestion
  "x": "t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJ
DnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2",
  "d": "IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"
```

emlun replied:
`x` and `d` are consistent in my testing.

selfissued replied:
I agree with @emlun's suggestion to use the same values for the COSE keys as were used for the corresponding JOSE keys.

dwaite replied:
I would suggest doing this for the next version, and combining the two appendices so that the JWK and CWK are next to one another.

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -210,75 +208,67 @@ E'(GF(p^8))
-  "y": "Bk0BZG3RijbTl9YqzuAZZHuxj4eNxKBF4TD_6KdxmCxgAri0Bx46xtcMqqKgqUmd
-DRC65Jv08EkVBojJkraUVEkxVhixgYS5hdfRfYZWaisT4M_NYoOFZFP4_5Jg05_p",
-  "d": "1jeo0uurZvaIC82YUMNLOHalfvhabbMHtLeVaI0tNhs"
+  "d": "TBD THIS IS NOW WRONG. REPLACE IT. VaI0tNhs"
```
```suggestion
  "x": "rgPyS3U6AR_cOuRC1LSxNmw5HfLlAurrRdFGjwyrZihZyeIoqMXQhr7aGM0eo1iS
E2WGstFUjZShhfoF2mK1D61CXFyOAoM-Qt6e7NZjsN9LKZNpGNHnFttuMrlBvk6v",
  "d": "O4x7NVtBvaLb-SZxGe3CKzl7obmSD47aLmReKdKQ_v8"
```

emlun replied:
`x` and `d` are consistent in my testing.

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -462,24 +422,21 @@ Below is the above CBOR rendered in diagnostic view.
-C6E754E865A9F01B3ED511D86673A6D8109DC94D0840FF539FA2D175B2DEA6BD3484D91A
-73424AD4C4ACC9CC748FB5280477A5C187DC6D749D0DE00BEEE1A206',
-  -4: h'2A9CFF87262CE1B093032446AC6911C4B947305DEEA7631E23D1A27C46BD9F62'
+  -4: h'TBD THIS IS NOW WRONG. REPLACE IT. 7305DEEA7631E23D1A27C46BD9F62'
```
```suggestion
  -2: h'b03ea8b181c4a4d3c5927a9ae3ad98587566b0dba9008240f5d54db7b19ef15c09
2e8baf88394a3fb52f063a6f56488a18e58f81d5ee1bf6c5146adcb1f43d746791322ee275
b022de06ed8b3f8ab56e53a1f6943662b3df528a5640a03eec6a',
  -4: h'3fe7a731678b27201792c20fb11fd91130edc405b8647cfd1ce4598e7186680d'
```

emlun replied:
`x` and `d` are consistent in my testing. Would it make sense for this to be the same key as the corresponding JWK example, though?

dwaite commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -434,20 +397,17 @@ Below is the above CBOR rendered in diagnostic view.
-77622EF08D0EBB092B5F7C7327A2352F06ABEFF3E0D8959F1F00276A96B7F02082798D84
-097CAE47119A14E450F7D93FB99FD597A10C0FF8C0563A25C2967455',
-  -4: h'890EC57AACF4D641C4B619A78774F9B814A50F5858AAFD56EC038739E96F834F'
+  -4: h'TBD THIS IS NOW WRONG. REPLACE IT. 50F5858AAFD56EC038739E96F834F'
```
```suggestion
    -2: h'89bf2eb011792886dc332582aaa7b2f2cd0e0428b3ed9b0f7fdfb3db1cb6c785
ccc3a2db601a3b9fc3b27471f3705cba15be70d37d81bd7e244af2b177c26892fb5770
def9c321d7055b2635cf74b9ce8e300c3d052304a03bf42002fd395b22',
  -4: h'60a16f19075486bf6fda27796b91c9149992d0ffa623b0be8c5e1b413e76d481'
```

emlun replied:
`x` and `d` are consistent in my testing. Would it make sense for this to be the same key as the corresponding JWK example, though?

selfissued committed:
Applied DW's suggestion

Co-authored-by: David Waite <david@alkaline-solutions.com>

selfissued committed:
Applied DW's suggestion

Co-authored-by: David Waite <david@alkaline-solutions.com>

selfissued committed:
DW's G2 examples

emlun commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -210,75 +208,67 @@ E'(GF(p^8))
-  "d": "hR6HfxlTwcjMGST5wYnkGiJvuVnpUPbvXSGsvwjJhUM"
+  "x": "t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJ
+DnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2",
+  "d": "IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"
```
This `d` value is big-endian, but the above definition of `"d"` states that its "value MUST contain the little-endian representation of the private key". I think the definition should be big-endian instead (which would then agree with this example).

We can check the endianness by checking for consistency with the public key `x`. First using [@noble/curves](https://github.com/paulmillr/noble-curves) to decode the public key and confirm consistency:

```
$ node
> const { bls12_381} = require("@noble/curves/bls12-381")
> const b64u = require("base64url")
> d = bls12_381.fields.Fr.fromBytes(b64u.toBuffer("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"))
15386317252953966243068698475713215110548824422523943950150305918933148006670n
> x = bls12_381.G2.CURVE.fromBytes(b64u.toBuffer("t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJDnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2"))
{
  x: {
    c0: 2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198n,
    c1: 3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385n
  },
  y: {
    c0: 1091225907871541971099156319884496768994573942435479350626538810779395867698340173817993335860569537230377198516854n,
    c1: 2595333569697813473526172873449378383222805423158275723970303644390236295775373772592557270513291928304374795997738n
  }
}
> bls12_381.G2.ProjectivePoint.fromPrivateKey(d)
Point {
  px: {
    c0: 2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198n,
    c1: 3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385n
  },
  py: {
    c0: 1091225907871541971099156319884496768994573942435479350626538810779395867698340173817993335860569537230377198516854n,
    c1: 2595333569697813473526172873449378383222805423158275723970303644390236295775373772592557270513291928304374795997738n
  },
  pz: { c0: 1n, c1: 0n }
}
```

Then using Python to confirm the endianness:

```
$ python
>>> import base64
>>> int.from_bytes(base64.urlsafe_b64decode("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4="), 'big')
15386317252953966243068698475713215110548824422523943950150305918933148006670
>>> int.from_bytes(base64.urlsafe_b64decode("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4="), 'little')
6547402858493071900195431625252553980600119178862449015842830395755105616930
```

So we see that at least the [@noble/curves](https://github.com/paulmillr/noble-curves) implementation uses big-endian encoding of `d`. I haven't compared with other implementations yet, but big-endian is more consistent with the point encoding, since the coordinates (`x.c1` and `x.c0`) are also encoded in big-endian:

```
$ python
>>> x = base64.urlsafe_b64decode("t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJDnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2")
>>> int.from_bytes(bytes([x[0] & 0x1f, *x[1:48]]), 'big')
3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385
>>> int.from_bytes(x[48:], 'big')
2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198
```

So I think the definition of `"d"` should change to use big-endian encoding there too.

selfissued replied:
Changed the description of "d" to big-endian.

emlun reviewed (dismissed):
The definition of `"d"` still requires the value to be little-endian, but I think big-endian is more appropriate since that is the byte order used internally in the encoding of the coordinates of public key points. See inline comment for details.

selfissued committed:
Change d to big-endian

emlun commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -71,21 +71,19 @@ The following definitions apply to the pairing-friendly elliptic curves known as
-- The parameter "y" MUST be present whose value represents the y coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
+- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.
+- The parameter "d" MUST be present for private key representations whose value MUST contain the big-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
```
Should we also define `"d"` to be of fixed length? Similar to [RFC 7518](https://www.rfc-editor.org/rfc/rfc7518.html#section-6.2.2.1):

>The length of this octet string MUST be ceiling(log-base-2(n)/8) octets (where n is the order of the curve).

I've seen enough interop issues with presence vs. absence of leading zeroes etc (mostly in ASN.1, though) even in my short career. :slightly_smiling_face:

selfissued replied:
I agree that these should be fixed length, including leading zeros. What sizes are they for each curve?

emlun replied:
32 bytes for [BLS12-381](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-11.html#name-bls-curves-for-the-128-bit-):

```
r: 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001
```

and 65 bytes for [BLS48-581](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-11.html#name-for-256-bit-security):

```
r: 0x2386f8a925e2885e233a9ccc1615c0d6c635387a3f0b3cbe003fad6bc972c2e6e741969d34c4c92016a85c7cd0562303c4ccbe599467c24da118a5fe6fcd671c01
```

dwaite replied:
I can generate new (matching) BLS12381G2 examples if we want to change `d` to be big-endian and fixed length.

emlun replied:
The example `d`s are already big-endian as far as I could tell. :slightly_smiling_face:

emlun committed:
Fix mmark reference syntax

emlun committed:
Update sections 2.1 and 2.2.3 to match new point encoding

emlun committed:
Define point encoding for BLS48-581

emlun committed:
Require fixed-length private key

This has precedent in RFC 7518, section 6.2.2.1:
https://www.rfc-editor.org/rfc/rfc7518.html#section-6.2.2.1

And also indirectly in RFC 8037:
https://www.rfc-editor.org/rfc/rfc8037.html#section-2
since it implicitly delegates the private key encoding to RFC 8032 which also
uses a fixed-length encoding: https://www.rfc-editor.org/rfc/rfc8032#section-3.1

BasileiosKal commented on draft-ietf-cose-bls-key-representations.md:
```diff
@@ -71,21 +71,19 @@ The following definitions apply to the pairing-friendly elliptic curves known as
-- The parameter "x" MUST be present whose value represents the x coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "y" MUST be present whose value represents the y coordinate of the curve point for the public key. This value MUST be encoded using the procedure defined in [Section 2.1](#point-coordinates-encoding) and MUST be base64url encoded without padding as defined in [@!RFC7515] Appendix C.
-- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
+- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.
```
```suggestion
- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized public key curve point, as defined normatively in Appendix B of <xref target="BBS"/>.
```

Just a small nit to make clear that the encoded point is the public key

BasileiosKal reviewed (approved):
Just a nit. LGTM!

selfissued committed:
Merge pull request #33 from emlun/mbj-zcash-d-fixed-length

Require fixed-length private key

selfissued committed:
Merge branch 'mbj-zcash' into mbj-zcash-bls48581

selfissued committed:
Merge pull request #32 from emlun/mbj-zcash-bls48581

Define BLS48-581 encoding, update other sections to match
